python/ray/experimental/data/dataset.py

```diff
@@ -465,14 +588,41 @@ def to_tf(self, **todo) -> "ray.util.sgd.tf.TFMLDataset":
         raise NotImplementedError  # P1
 
     def to_dask(self) -> "dask.DataFrame":
-        """Convert this dataset into a Dask dataframe.
+        """Convert this dataset into a Dask DataFrame.
 
         Time complexity: O(1)
 
         Returns:
-            A Dask dataframe created from this dataset.
+            A Dask DataFrame created from this dataset.
         """
-        raise NotImplementedError  # P1
+        import dask
+        import dask.dataframe as dd
+        from ray.util.client.common import ClientObjectRef
+
+        def _block_to_df(block):
+            return block._table.to_pandas()
+
+        _remote_block_to_df = ray.remote(_block_to_df)
+
+        @dask.delayed
+        def block_to_df(block):
+            if isinstance(block, (ray.ObjectRef, ClientObjectRef)):
+                # The block Ray future hasn't been resolved to a concrete
+                # value, so the end-user isn't using Dask-on-Ray. We therefore
+                # launch a converting Ray task and block here to get the
+                # underlying dataframe.
+                df = ray.get(_remote_block_to_df.remote(block))
+            else:
+                # The block Ray future has been resolved, so the end-user is
+                # using Dask-on-Ray. We therefore directly convert the block to
+                # a dataframe since we're already running in a Ray task.
+                df = _block_to_df(block)
+
+            return df
+
+        # TODO(Clark): Give Dask a Pandas-esque schema via the Pyarrow schema,
+        # once that's implemented.
+        return dd.from_delayed([block_to_df(block) for block in self._blocks])
```

Comment:
Will this dataframe automatically use dask on Ray? We should either enable dask on ray for the df, or if not possible just enable it globally when you call .to_dask().

Reply:
We could set the returned dataframe's `__dask_scheduler__` attribute, although that's supposed to be defined by the implementer of a custom collection, not mutated at runtime by a Dask integration, so that seems like a bit of an antipattern. What do you think?

I also don't think that we should be transparently setting those kind of global configs for external libraries in this case, since the user may have their reason for not wanting to use the Dask-on-Ray scheduler and I don't think that we should box them in like that.

Reply:
I'll try out setting the `__dask_scheduler__` attribute.

Reply:
It looks like we'll have to set the Dask scheduler globally via the config (for now) due to some precomputation that Dask will do on these delayed objects in order to determine the DataFrame schema.

Reply:
Ok let's try that for now. We can always add an option to turn it off if users want, but most of the times running two different cluster schedulers in one program doesn't make much sense (probably an error).